File: L3/Problems.cpp
//
// Created by poloj on 17-07-2026.
//
#include<stdio.h>
int length(int arr[]);
int TwoSum(int arr[],int target,int ans[]);
void ProductOfArray(int arr[]);
int main(void) {
    int arr[]={2,7,11};
    int target=9;
    int ans[2];
    TwoSum(arr,target,ans);
    printf("%d %d",ans[0],ans[1]);
    //---------------------------------
    //failed one
    // int arr[]={1,2,3,4};
    // ProductOfArray(arr);
    return 0;
}
int length(int arr[]) {
    int count=0;
    while(arr[count]!=0) {
        count=count+1;

    }
    return count;
}
int TwoSum(int arr[],int target,int ans[]) {
    int n =length(arr);
    //printf("%d",n);
    int left=0;
    int right=n-1;
    while (left<right) {

        if (arr[left]+arr[right]==target) {
            ans[0]=left;
            ans[1]=right;
            break;
        }
        else if (arr[left]+arr[right]<target) {
            left=left+1;
        }
        else {
            right=right-1;
        }
    }
    if (ans[0]!=-1) {
        return ans[0],ans[1];
    }
    else {
        return -1;
    }

}
//failed in time complexity
void ProductOfArray(int arr[]){
    int n =length(arr);
    int ans[n-1];
    int left=0;
    int i=0,product=1;
    int count=0;
    while (left<n) {
        if (i==left) {
            i++;
            continue;
        }
        product=product*arr[i];
        i=i+1;
        count=count+1;
        if (count==n-2) {
           printf("%d,",product);
            ans[left]=product;
            left=left+1;
            product=1;
            count=0;
            i=0;
        }

    }
    // for (int i=0;i<n-1;i++) {
    //     printf("%d,",ans[i]);
    // }

}

File: L3/Python/Problems.py
def length(arr):
    count = 0
    for _ in arr:
        count += 1
    return count

def TwoSum(arr,target):
    n = length(arr)
    left=0
    right=n-1
    arr.sort()
    while(left<right):
        if(arr[left]+arr[right]==target):
            return [left,right]
        elif arr[left]+arr[right]<target:
            left=left+1
        else:
            right=right-1
    if ans!=-1:
        return ans
    else:
        return -1
# arr=[2,7,11]
# target=9
# ans=TwoSum(arr,target)
# print(ans)
def ProductOfArray(arr):
    n = length(arr)
    left=0
    product=1
    count=0
    i=0
    ans=[n]
    while left<n:
        if i == left:
            i=i+1
            continue
        product=product*arr[i]
        i=i+1
        count=count+1;
        if count==n-1:
            print(product,end=" ")
            ans.append(product)
            left=left+1
            count=0
            i=0
            product=1

    return ans
# arr=[1,2,3,4]
# ProductOfArray(arr)
def subArraySum(arr,target):
    n = length(arr)
    left=0
    i=0
    count=0
    right=left+1
    sum=0
    while(left<n-1):
        sum=sum+arr[i]
        if(sum==target):
            left=left+1
            count=count+1
            sum=0
        else:
            right=right+1
            i=i+1

    print(count)
arr= [1,1,1]
k=2
subArraySum(arr,k)
def longestSubString(str1):
    n = length(str1)
    ls = list(str1)
    un = set(ls)
    ans = ""
    for i in un:
        ans=ans+i
    if ans in str1:
        return ans

    # print(ls)
    # print(un)
    # print(ans)
# print(longestSubString("abcabcbb"))
